### PR TITLE
Woptim/update local packages

### PR DESCRIPTION
**Pull request by adrienbernede**
## Summary

- This PR is an improvement.
- It does the following:
  - Update local Spack packages based on current spack@develop version. Keep the logic that was only present locally when necessary to pass tests (see commit messages)
  - Modifies the CI to add necessary variants.

## Important notes

The raja package diff is hard to read because of the switch to CachedCMakePackages. Reviewers should compare the new version with https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/raja/package.py.

By default, "shared" is "True" in the upstream spack package. I set "shared" to "false" in the global variants settings in CI here:

By default, "tests" is "False" in the upstream spack package. I set "tests" to "true" in the global variants settings in CI here:

https://github.com/LLNL/RAJA/blob/eb97f73f836afa6d49491d3b1938a395880c9bfe/.gitlab/custom-jobs-and-variables.yml#L21
https://github.com/LLNL/RAJA/blob/eb97f73f836afa6d49491d3b1938a395880c9bfe/.gitlab/custom-jobs-and-variables.yml#L31
https://github.com/LLNL/RAJA/blob/eb97f73f836afa6d49491d3b1938a395880c9bfe/.gitlab/custom-jobs-and-variables.yml#L40

Relates to:
https://github.com/LLNL/Umpire/pull/793.
https://github.com/LLNL/CHAI/pull/218

### CODE REVIEW

**Comment by adrienbernede** [comment]
@reviewers:

I took the packages for spack@develop, ran the CI, and added the missing parts from the initial local packages to pass the tests.

It is possible that not all the diffs were ported to the new local packages. But since we are now passing the tests with packages much closer to the upstream packages, it seems undeniable that we should take this as a new start.

Also remember that the goal is to minimize the diffs with the upstream package. We should consider pushing the new local packages to Spack upstream.

This is part of the process of choosing a Spack reference for CI testing described here:
https://radiuss-shared-ci.readthedocs.io/en/latest/sphinx/user_guide/how_to.html#choose-a-spack-reference-commit-or-branch

**Comment by rhornung67** [comment]
Adding more reviewers since we will discuss Gitlab CI, including this at the project meeting tomorrow morning.

**Comment by rhornung67** [comment]
@adrienbernede there is an issue with the spack spec in one of the lassen xl+cuda builds that is causing the failure.

**Comment by adrienbernede** [comment]
@rhornung67 This should be fixed.

**Comment by rhornung67** [comment]
@adrienbernede can you still make the RAJA meeting tomorrow? We can discuss these and answer folks' questions before merging them.

**Comment by adrienbernede** [comment]
@rhornung67 Sure I can.

**Comment by adrienbernede** [comment]
I would ask for a review from either @trws or @davidbeckingsale. I pushed it to share, it is possible that it needs some work.

**Comment by rhornung67** [comment]
Also, we may need to hold off on merging this until after the release since the CUDA 11.0.2 build will fail with https://github.com/LLNL/RAJA/pull/1004. @mdavis36 reported that it needs CUDA 11.1.x

**Comment by adrienbernede** [comment]
@trws The goal of this PR is to get closer to the upstream version of the packages, and then hopefully push the changes to spack so that we don’t need the local packages anymore.
We can do it in 2 steps as described, or we could also directly push the necessary changes to Spack and remove the local packages in RAJA while updating Spack.
In any case, using the current upstream packages, the RAJA build would fail on LC machines, so an update of the packages is necessary.

**Comment by adrienbernede** [comment]
@rhornung67 I’ll have a closer look to the failure of `xl_16_1_1_12_gcc_8_3_1_cuda_11_0_2`. But I could definitely either allow it to fail or prevent it to run.

**Comment by rhornung67** [comment]
Thanks @adrienbernede. 

As an aside: Do you know if there is a way to have GItlab somehow coordinate with the LC job schedules so we don't have so many jobs fail when they timeout waiting in the queue before they have a chance to run?

**Comment by adrienbernede** [comment]
Conflict resolution report:
- The new version of the raja package was already defining CMAKE_CUDA_ARCHITECTURE (beside CUDA_ARCH, which may be useless).
- The CMake executable used in the build-and-test script is now derived from the host-config (cmake config cache) file, instead of loading cmake modules. If this is expected to fail is situations, please educate me.

**Comment by adrienbernede** [comment]
Note: the new CMake requirements had not been added to the local RAJA package. This was not detected because the build and test script was using an external CMake different than the one use by Spack to build the dependencies.

I fixed that and updated the RAJA package.

@davidbeckingsale If you are available, I’d like your input on this: I am willing to update the local packages of Umpire and CHAI as well, which would be useful, notably for CHAI CI.

File: scripts/spack_packages/camp/package.py
Comment (rhornung67) [inline scripts/spack_packages/camp/package.py:32]: Add version v2022.10.0, which is what RAJA develop is using.

**Review by rhornung67 — COMMENTED** [review]
A couple of spack package version updates.

**Comment by adrienbernede** [comment]
@rhornung67 will do.

File: scripts/spack_packages/camp/package.py
Comment (adrienbernede) [reply scripts/spack_packages/camp/package.py:33]: Done

File: scripts/spack_packages/raja/package.py
Comment (adrienbernede) [reply scripts/spack_packages/raja/package.py:68]: Done

**Comment by rhornung67** [comment]
@adrienbernede this looks good to me. I would like @davidbeckingsale to review.

**Comment by adrienbernede** [comment]
@rhornung67 OK.

For the records, I am trying to sync things with Umpire and CHAI.

Illustration of the challenges:
- Since CHAI@develop needs 2022.10 releases, we need the updated packages locally in CHAI.
- the 2022.10 releases require BLT 5.2.0 which is not present in spack@v0.18.1. This means we need a local BLT package for BLT... again. Except if we push all our new packages to a Spack PR and get it merged.

Suggested path forward:
1. Make things work _without updating Spack yet_.
2. Push the new packages to Spack in a joint PR (RAJA, Umpire, CHAI, camp?).
3. Update Spack in our project and remove identical local packages.

**Review by davidbeckingsale — APPROVED** [review]

**Review by rhornung67 — APPROVED** [review]
Looks good to me. Thank you @adrienbernede

**Comment by rhornung67** [comment]
@davidbeckingsale should I merge this? I don't foresee any issues with it.

**Comment by davidbeckingsale** [comment]
> @davidbeckingsale should I merge this? I don't foresee any issues with it.

Yup, sounds good!

**Pull request merged**
